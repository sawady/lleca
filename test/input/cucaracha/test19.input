
fun f(x : Int, y : Int) {
  putNum(x - y)
  putChar(10)
}

fun main() {
  putNum(100 - 58)
  putChar(10)
  f(100, 58)

  a := 100
  b := 58
  putNum(a - b)
  putChar(10)
  f(a, b)

  a := a + 1
  b := b - 1
  a := 2 * a
  b := 2 * b
  putNum(a - b)
  putChar(10)
  f(a, b)

  a := a + 1
  b := b - 1
  a := 2 * a
  b := 2 * b
  putNum(a - b)
  putChar(10)
  f(a, b)

  a := a + 1
  b := b - 1
  a := 2 * a
  b := 2 * b
  putNum(a - b)
  putChar(10)
  f(a, b)

  a := a + 1
  b := b - 1
  a := 2 * a
  b := 2 * b
  putNum(a - b)
  putChar(10)
  f(a, b)

  a := a + 1
  b := b - 1
  a := 2 * a
  b := 2 * b
  putNum(a - b)
  putChar(10)
  f(a, b)

  putNum(a) putChar(10)
  putNum(b) putChar(10)
}

