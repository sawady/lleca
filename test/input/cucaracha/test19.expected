Cons(
  Function(
    f,
    Cons(
      Parameter(
        x,
        Int
      ),
      Cons(
        Parameter(
          y,
          Int
        ),
        Nil
      )
    ),
    Unit,
    Cons(
      StmtCall(
        putNum,
        Cons(
          ExprSub(
            ExprVar(
              x
            ),
            ExprVar(
              y
            )
          ),
          Nil
        )
      ),
      Cons(
        StmtCall(
          putChar,
          Cons(
            ExprConstNum(
              10
            ),
            Nil
          )
        ),
        Nil
      )
    )
  ),
  Cons(
    Function(
      main,
      Nil,
      Unit,
      Cons(
        StmtCall(
          putNum,
          Cons(
            ExprSub(
              ExprConstNum(
                100
              ),
              ExprConstNum(
                58
              )
            ),
            Nil
          )
        ),
        Cons(
          StmtCall(
            putChar,
            Cons(
              ExprConstNum(
                10
              ),
              Nil
            )
          ),
          Cons(
            StmtCall(
              f,
              Cons(
                ExprConstNum(
                  100
                ),
                Cons(
                  ExprConstNum(
                    58
                  ),
                  Nil
                )
              )
            ),
            Cons(
              StmtAssign(
                a,
                ExprConstNum(
                  100
                )
              ),
              Cons(
                StmtAssign(
                  b,
                  ExprConstNum(
                    58
                  )
                ),
                Cons(
                  StmtCall(
                    putNum,
                    Cons(
                      ExprSub(
                        ExprVar(
                          a
                        ),
                        ExprVar(
                          b
                        )
                      ),
                      Nil
                    )
                  ),
                  Cons(
                    StmtCall(
                      putChar,
                      Cons(
                        ExprConstNum(
                          10
                        ),
                        Nil
                      )
                    ),
                    Cons(
                      StmtCall(
                        f,
                        Cons(
                          ExprVar(
                            a
                          ),
                          Cons(
                            ExprVar(
                              b
                            ),
                            Nil
                          )
                        )
                      ),
                      Cons(
                        StmtAssign(
                          a,
                          ExprAdd(
                            ExprVar(
                              a
                            ),
                            ExprConstNum(
                              1
                            )
                          )
                        ),
                        Cons(
                          StmtAssign(
                            b,
                            ExprSub(
                              ExprVar(
                                b
                              ),
                              ExprConstNum(
                                1
                              )
                            )
                          ),
                          Cons(
                            StmtAssign(
                              a,
                              ExprMul(
                                ExprConstNum(
                                  2
                                ),
                                ExprVar(
                                  a
                                )
                              )
                            ),
                            Cons(
                              StmtAssign(
                                b,
                                ExprMul(
                                  ExprConstNum(
                                    2
                                  ),
                                  ExprVar(
                                    b
                                  )
                                )
                              ),
                              Cons(
                                StmtCall(
                                  putNum,
                                  Cons(
                                    ExprSub(
                                      ExprVar(
                                        a
                                      ),
                                      ExprVar(
                                        b
                                      )
                                    ),
                                    Nil
                                  )
                                ),
                                Cons(
                                  StmtCall(
                                    putChar,
                                    Cons(
                                      ExprConstNum(
                                        10
                                      ),
                                      Nil
                                    )
                                  ),
                                  Cons(
                                    StmtCall(
                                      f,
                                      Cons(
                                        ExprVar(
                                          a
                                        ),
                                        Cons(
                                          ExprVar(
                                            b
                                          ),
                                          Nil
                                        )
                                      )
                                    ),
                                    Cons(
                                      StmtAssign(
                                        a,
                                        ExprAdd(
                                          ExprVar(
                                            a
                                          ),
                                          ExprConstNum(
                                            1
                                          )
                                        )
                                      ),
                                      Cons(
                                        StmtAssign(
                                          b,
                                          ExprSub(
                                            ExprVar(
                                              b
                                            ),
                                            ExprConstNum(
                                              1
                                            )
                                          )
                                        ),
                                        Cons(
                                          StmtAssign(
                                            a,
                                            ExprMul(
                                              ExprConstNum(
                                                2
                                              ),
                                              ExprVar(
                                                a
                                              )
                                            )
                                          ),
                                          Cons(
                                            StmtAssign(
                                              b,
                                              ExprMul(
                                                ExprConstNum(
                                                  2
                                                ),
                                                ExprVar(
                                                  b
                                                )
                                              )
                                            ),
                                            Cons(
                                              StmtCall(
                                                putNum,
                                                Cons(
                                                  ExprSub(
                                                    ExprVar(
                                                      a
                                                    ),
                                                    ExprVar(
                                                      b
                                                    )
                                                  ),
                                                  Nil
                                                )
                                              ),
                                              Cons(
                                                StmtCall(
                                                  putChar,
                                                  Cons(
                                                    ExprConstNum(
                                                      10
                                                    ),
                                                    Nil
                                                  )
                                                ),
                                                Cons(
                                                  StmtCall(
                                                    f,
                                                    Cons(
                                                      ExprVar(
                                                        a
                                                      ),
                                                      Cons(
                                                        ExprVar(
                                                          b
                                                        ),
                                                        Nil
                                                      )
                                                    )
                                                  ),
                                                  Cons(
                                                    StmtAssign(
                                                      a,
                                                      ExprAdd(
                                                        ExprVar(
                                                          a
                                                        ),
                                                        ExprConstNum(
                                                          1
                                                        )
                                                      )
                                                    ),
                                                    Cons(
                                                      StmtAssign(
                                                        b,
                                                        ExprSub(
                                                          ExprVar(
                                                            b
                                                          ),
                                                          ExprConstNum(
                                                            1
                                                          )
                                                        )
                                                      ),
                                                      Cons(
                                                        StmtAssign(
                                                          a,
                                                          ExprMul(
                                                            ExprConstNum(
                                                              2
                                                            ),
                                                            ExprVar(
                                                              a
                                                            )
                                                          )
                                                        ),
                                                        Cons(
                                                          StmtAssign(
                                                            b,
                                                            ExprMul(
                                                              ExprConstNum(
                                                                2
                                                              ),
                                                              ExprVar(
                                                                b
                                                              )
                                                            )
                                                          ),
                                                          Cons(
                                                            StmtCall(
                                                              putNum,
                                                              Cons(
                                                                ExprSub(
                                                                  ExprVar(
                                                                    a
                                                                  ),
                                                                  ExprVar(
                                                                    b
                                                                  )
                                                                ),
                                                                Nil
                                                              )
                                                            ),
                                                            Cons(
                                                              StmtCall(
                                                                putChar,
                                                                Cons(
                                                                  ExprConstNum(
                                                                    10
                                                                  ),
                                                                  Nil
                                                                )
                                                              ),
                                                              Cons(
                                                                StmtCall(
                                                                  f,
                                                                  Cons(
                                                                    ExprVar(
                                                                      a
                                                                    ),
                                                                    Cons(
                                                                      ExprVar(
                                                                        b
                                                                      ),
                                                                      Nil
                                                                    )
                                                                  )
                                                                ),
                                                                Cons(
                                                                  StmtAssign(
                                                                    a,
                                                                    ExprAdd(
                                                                      ExprVar(
                                                                        a
                                                                      ),
                                                                      ExprConstNum(
                                                                        1
                                                                      )
                                                                    )
                                                                  ),
                                                                  Cons(
                                                                    StmtAssign(
                                                                      b,
                                                                      ExprSub(
                                                                        ExprVar(
                                                                          b
                                                                        ),
                                                                        ExprConstNum(
                                                                          1
                                                                        )
                                                                      )
                                                                    ),
                                                                    Cons(
                                                                      StmtAssign(
                                                                        a,
                                                                        ExprMul(
                                                                          ExprConstNum(
                                                                            2
                                                                          ),
                                                                          ExprVar(
                                                                            a
                                                                          )
                                                                        )
                                                                      ),
                                                                      Cons(
                                                                        StmtAssign(
                                                                          b,
                                                                          ExprMul(
                                                                            ExprConstNum(
                                                                              2
                                                                            ),
                                                                            ExprVar(
                                                                              b
                                                                            )
                                                                          )
                                                                        ),
                                                                        Cons(
                                                                          StmtCall(
                                                                            putNum,
                                                                            Cons(
                                                                              ExprSub(
                                                                                ExprVar(
                                                                                  a
                                                                                ),
                                                                                ExprVar(
                                                                                  b
                                                                                )
                                                                              ),
                                                                              Nil
                                                                            )
                                                                          ),
                                                                          Cons(
                                                                            StmtCall(
                                                                              putChar,
                                                                              Cons(
                                                                                ExprConstNum(
                                                                                  10
                                                                                ),
                                                                                Nil
                                                                              )
                                                                            ),
                                                                            Cons(
                                                                              StmtCall(
                                                                                f,
                                                                                Cons(
                                                                                  ExprVar(
                                                                                    a
                                                                                  ),
                                                                                  Cons(
                                                                                    ExprVar(
                                                                                      b
                                                                                    ),
                                                                                    Nil
                                                                                  )
                                                                                )
                                                                              ),
                                                                              Cons(
                                                                                StmtAssign(
                                                                                  a,
                                                                                  ExprAdd(
                                                                                    ExprVar(
                                                                                      a
                                                                                    ),
                                                                                    ExprConstNum(
                                                                                      1
                                                                                    )
                                                                                  )
                                                                                ),
                                                                                Cons(
                                                                                  StmtAssign(
                                                                                    b,
                                                                                    ExprSub(
                                                                                      ExprVar(
                                                                                        b
                                                                                      ),
                                                                                      ExprConstNum(
                                                                                        1
                                                                                      )
                                                                                    )
                                                                                  ),
                                                                                  Cons(
                                                                                    StmtAssign(
                                                                                      a,
                                                                                      ExprMul(
                                                                                        ExprConstNum(
                                                                                          2
                                                                                        ),
                                                                                        ExprVar(
                                                                                          a
                                                                                        )
                                                                                      )
                                                                                    ),
                                                                                    Cons(
                                                                                      StmtAssign(
                                                                                        b,
                                                                                        ExprMul(
                                                                                          ExprConstNum(
                                                                                            2
                                                                                          ),
                                                                                          ExprVar(
                                                                                            b
                                                                                          )
                                                                                        )
                                                                                      ),
                                                                                      Cons(
                                                                                        StmtCall(
                                                                                          putNum,
                                                                                          Cons(
                                                                                            ExprSub(
                                                                                              ExprVar(
                                                                                                a
                                                                                              ),
                                                                                              ExprVar(
                                                                                                b
                                                                                              )
                                                                                            ),
                                                                                            Nil
                                                                                          )
                                                                                        ),
                                                                                        Cons(
                                                                                          StmtCall(
                                                                                            putChar,
                                                                                            Cons(
                                                                                              ExprConstNum(
                                                                                                10
                                                                                              ),
                                                                                              Nil
                                                                                            )
                                                                                          ),
                                                                                          Cons(
                                                                                            StmtCall(
                                                                                              f,
                                                                                              Cons(
                                                                                                ExprVar(
                                                                                                  a
                                                                                                ),
                                                                                                Cons(
                                                                                                  ExprVar(
                                                                                                    b
                                                                                                  ),
                                                                                                  Nil
                                                                                                )
                                                                                              )
                                                                                            ),
                                                                                            Cons(
                                                                                              StmtCall(
                                                                                                putNum,
                                                                                                Cons(
                                                                                                  ExprVar(
                                                                                                    a
                                                                                                  ),
                                                                                                  Nil
                                                                                                )
                                                                                              ),
                                                                                              Cons(
                                                                                                StmtCall(
                                                                                                  putChar,
                                                                                                  Cons(
                                                                                                    ExprConstNum(
                                                                                                      10
                                                                                                    ),
                                                                                                    Nil
                                                                                                  )
                                                                                                ),
                                                                                                Cons(
                                                                                                  StmtCall(
                                                                                                    putNum,
                                                                                                    Cons(
                                                                                                      ExprVar(
                                                                                                        b
                                                                                                      ),
                                                                                                      Nil
                                                                                                    )
                                                                                                  ),
                                                                                                  Cons(
                                                                                                    StmtCall(
                                                                                                      putChar,
                                                                                                      Cons(
                                                                                                        ExprConstNum(
                                                                                                          10
                                                                                                        ),
                                                                                                        Nil
                                                                                                      )
                                                                                                    ),
                                                                                                    Nil
                                                                                                  )
                                                                                                )
                                                                                              )
                                                                                            )
                                                                                          )
                                                                                        )
                                                                                      )
                                                                                    )
                                                                                  )
                                                                                )
                                                                              )
                                                                            )
                                                                          )
                                                                        )
                                                                      )
                                                                    )
                                                                  )
                                                                )
                                                              )
                                                            )
                                                          )
                                                        )
                                                      )
                                                    )
                                                  )
                                                )
                                              )
                                            )
                                          )
                                        )
                                      )
                                    )
                                  )
                                )
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    ),
    Nil
  )
)
